### PR TITLE
PCRM-3852-Test-Coverage

### PR DESCRIPTION
**Pull request by HarveyNHS**
Test Coverage for PCRM-3852

### CODE REVIEW

File: test/integrationtests/specs/lr_only_on_pds_process.spec [hunk 1]
```diff
@@ -0,0 +1,34 @@
+# Test OnlyOnPDS.csv process
+----------------------------
+Tags: wip
+
+## Test to ensure if any outstanding records that are on PDS API data but are not on GP flat file can be added to OnlyOnPDS.csv output file
+-------------------------------------------------------------------------------------------------------------------------------------------
+
```
Comment (premnhs) [inline test/integrationtests/specs/lr_only_on_pds_process.spec:7]: remove this empty line

File: test/integrationtests/specs/lr_only_on_pds_process.spec [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+* setup step: connect to s3 buckets LR-20 and upload MESH data
+* setup step: connect to s3 bucket mock pds api and upload csv file that contains the missing records
+* upload a GP flat file that has missing PDS data
+* connect to lr-03 dynamodb and get the latest JobId for a gppractice file
+* trigger lr09 and ensure scheduled checked successfully completed
+* check LR-13 contains only on pds output file
+
+## Test to ensure that if the MESH data is late the process for producing a OnlyOnPDS.csv can be created once a GP flat file has been processed again and MESH has been split
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+
```
Comment (premnhs) [inline test/integrationtests/specs/lr_only_on_pds_process.spec:17]: same here

File: test/integrationtests/step_impl/lr_only_on_pds_steps.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+# On github
+access_key = os.getenv("AWS_PUBLIC_KEY")
+secret_key = os.getenv("AWS_PRIVATE_KEY")
+dev = boto3.session.Session(access_key, secret_key)
+
+
+REGION_NAME = "eu-west-2"
+
+AWS_RESOURCE = get_aws_resources()
+test_datetime = get_datetime_now()
```
Comment (jamesnunnnhs) [inline test/integrationtests/step_impl/lr_only_on_pds_steps.py:24]: The convention is that all your top level constants should be capitalised - it makes it much easier to see where it is defined

File: test/integrationtests/step_impl/lr_only_on_pds_steps.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+
+@step("check LR-13 contains only on pds output file")
+def check_lr_13_contains_only_on_pds_output_file():
+    job_id_prefix = get_latest_jobid()
+
+    s3_client = dev.client("s3", REGION_NAME)
+    response = s3_client.list_objects_v2(Bucket=LR_13_BUCKET, Prefix=job_id_prefix)
+
+    if response.get("Contents") != None:
```
Comment (jamesnunnnhs) [inline test/integrationtests/step_impl/lr_only_on_pds_steps.py:128]: Doesn't this mask a failing condition if the file doesn't exist? Maybe better to remove this check?

Please check the other patterns like this, and also `if "Contents" in lr_22_response:` is the same check.

**Review by jamesnunnnhs — CHANGES_REQUESTED** [review]

File: test/integrationtests/step_impl/lr_only_on_pds_steps.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        for obj in response.get("Contents"):
+            print("Found CSV File", obj["Key"])
+            s3_client.get_object(Bucket=LR_13_BUCKET, Key=obj.get("Key"))
+
+        expected_string = obj.get("Key")
+        filename, ext = os.path.splitext(expected_string)
+        elements = "-".join(filename.split("-")[0:6])
+
+        assert elements == f"{job_id_prefix}/A76543-OnlyOnPDS"
+        assert ext == ".csv"
```
Comment (premnhs) [inline test/integrationtests/step_impl/lr_only_on_pds_steps.py:138]: Is it similar to the steps in `lr_10_12_13_onlyon_pds.py` ( step : "connect to lr-13 and check output csv file content and file format as expected") 

Why cant this step be used ?

File: test/integrationtests/specs/lr_only_on_pds_process.spec [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+* connect to lr-03 dynamodb and get the latest JobId for a gppractice file
+* trigger lr09 and ensure scheduled checked successfully completed
+* check LR-13 is empty
+* setup step: connect to s3 buckets LR-20 and upload MESH data
+* get InFlight table item count
+* upload a GP flat file that has missing PDS data
+* connect to lr-03 dynamodb and get the latest JobId for a gppractice file
+* trigger lr09 and ensure scheduled checked successfully completed
+* check LR-13 contains only on pds output file
+
```
Comment (premnhs) [inline test/integrationtests/specs/lr_only_on_pds_process.spec:29]: Can these tests be added as part of this spec 
`  lr_10_12_13_lambda_onlyon_pds.spec`

File: main.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
+import os
```
Comment (josh-scargill-nhsd) [inline main.py:1]: This file shouldn't be committed in

**Review by josh-scargill-nhsd — CHANGES_REQUESTED** [review]

File: test/integrationtests/step_impl/lr_only_on_pds_steps.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+
+ROOT = os.path.dirname(os.path.abspath(__file__))
+DATA = os.path.join(ROOT, "data")
+MOCK_PDS_DATA = AWS_RESOURCE["mock_pds_data"]["value"]
+
+
+@step("setup step: connect to s3 buckets LR-20 and upload MESH data")
+def setup_step_connect_to_s3_bucket_lr_20_upload_data():
+    dps_data = "dps_data.csv"
+
```
Comment (premnhs) [inline test/integrationtests/step_impl/lr_only_on_pds_steps.py:40]: why there is a line space here ?

File: test/integrationtests/step_impl/lr_only_on_pds_steps.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+    upload_dps_data = os.path.join(DATA, "OnlyOnPDS/" + dps_data)
+
```
Comment (premnhs) [inline test/integrationtests/step_impl/lr_only_on_pds_steps.py:42]: same here ?

File: test/integrationtests/step_impl/lr_only_on_pds_steps.py [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+    s3 = dev.client("s3", REGION_NAME)
+
```
Comment (premnhs) [inline test/integrationtests/step_impl/lr_only_on_pds_steps.py:44]: same here

**Review by jamesnunnnhs — APPROVED** [review]

**Review by premnhs — APPROVED** [review]
LGTM

**Review by josh-scargill-nhsd — APPROVED** [review]

**Pull request merged**